func_interface.go

```diff
@@ -0,0 +1,27 @@
+package taskgraph
+
+// We need some interface to define function and how we optimize them.
+type Parameter interface {
+	GetParam(index uint64) float32
+	SetParam(index uint64, value float32)
+}
+
+type Gradient interface {
+	GetValue() float32
+	GetGradient(index uint64) float32
+	SetValue(value float32)
+	SetGradient(index uint64, grad float32)
+}
+
+// There are many different ways one can optimize a function, but the
+// most effective ones need gradient.
+type Function interface {
```

Comment:
the implementation is free to do what ever it wants, but they should be a function.